test/persistent_receiver_test.go

```diff
@@ -1696,19 +1694,46 @@ var _ = Describe("PersistentReceiver", func() {
 
 				It("handles unsolicited terminations while already terminating", func() {
 					blocker := make(chan struct{})
+                    signal := make(chan bool)
 					messageChannel := make(chan message.InboundMessage, numMessages+1)
+                    waitToReceiveMessages := 30 * time.Second // This timer needs to be big enough to give us a very
+                                                              // high chance of both receiving the messages and getting
+                                                              // past the state change in Terminate. It may need to be
+                                                              // updated in the future since this only mitigates the
+                                                              // race condition and doesn't eliminate it.
 					messageReceiver.ReceiveAsync(func(inboundMessage message.InboundMessage) {
+                        loop:
+                            for timeout := time.After(waitToReceiveMessages); ; {
+                                select {
```

Comment:
@TrentDaniel, please what is the purpose for this loop and timeout? The message callback will be executed by the `ReceiveAsync()` when an actual message is received from the broker.

Reply:
The loop will exit after the timeout period has expired, regardless of whether or not the receiver is running. The loop will also exit before the timeout expires if it finds that the receiver is no longer running. In either case, the signal channel will be written to and this function will exit. This will unblock the main goroutine that is running test but waiting to read from the signal channel, and will also allow the callback function to complete and its goroutine to exit. This polling loop is how  to avoid the hang that you were concerned about.

Reply:
Okay, thanks for the clarification. I think it is okay.

Reply:
I am curious to know what @cjwmorgan-sol thinks about this for loop inside the ReceiveAysn() method ?